flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/KMeans.java

```diff
@@ -146,7 +145,7 @@ public IterationBodyResult process(
             DataStream<DenseVector> points = dataStreams.get(0);
 
             DataStream<Integer> terminationCriteria =
-                    centroids.flatMap(new TerminateOnMaxIterationNum<>(maxIterationNum));
+                    centroids.map(x -> 0.).flatMap(new TerminationCriteria(maxIterationNum));
```

Comment:
Would it be better to still allow `TerminationCriteria` to accept arbitrary input type so that the caller code does not have to explicitly convert input type to Integer?

Reply:
During ML training, a common termination criteria is to terminate when number of iterations exceeds the specified max number of iterations, or the loss is smaller than a given tolerance (a tolerance is a `double`). 

If we want to allow termination on a data stream with arbitrary input type, we probably need to add a new class (i.e., `TerminateOnMaxIterationNum` here). 

IMO, maintaining two utility classes for termination may confuse the developers. What do you think?

Reply:
IMO, for users who wants to termination iteration based on round number, asking them to map input to double-typed values seems unnecessary and might confuse users.

And if we have two separate classes such as `TerminateOnMaxIterationNum ` and `TerminateOnToleranceThreshold`, given that users already have concepts of these two types of termination pattern, the class names seem to be pretty self-explanatory and intuitive.

What do you think? Or maybe we can wait for @gaoyunhaii comment?

Reply:
Having separate classess for different termination conditions is fine for me and also may not confuse users. Then we are going to have the following three classes:
- `TerminateOnMaxIterationNum`
- `TerminateOnToleranceThreshold`
- `TerminateOnMaxIterationNumOrToleranceThreshold

Do you think it is a viable solution?

Reply:
How about we have `TerminateOnMaxIter` and `TerminateOnMaxIterOrTol`?